### DISCUSSION TITLE
Change thread switch interval?

### DISCUSSION THREAD

**Discussion by puzzledsab** [sabnzbd/sabnzbd#2454] [General]
The thread switch interval (the time one thread will keep running if it keeps the GIL) is 0.005 seconds by default. I'm wondering if that may be a bit long? Most of the threads will only run for a very short time before sleeping or doing a system call which will release the thread. They may benefit from being called more frequently so that they can start a new decode or save another part. See the Convoy effect: https://bugs.python.org/issue7946

Opinions, @mnightingale @Safihre?

Get/set:
```
import sys
print(sys.getswitchinterval())
sys.setswitchinterval(0.001)
print(sys.getswitchinterval())
```

**Comment by mnightingale** [comment]
I briefly tried this a while ago, I couldn't see any difference but I didn't particularly profile it might cause higher load though with more context switching.

I did make a note to investigate https://github.com/chrisjbillington/gil_load, linux only and possibly untested on later Python versions to try and see how much time we are actually waiting for the GIL and hence if changing the switchinterval would have any affect.

As I see it there are only really 3 main time consuming threads; assembler, decoder and downloader.
I'd expect they all release the GIL for their time consuming parts (socket selects/IO, file I/O and decoding), so I have my doubts the GIL is actually currently a bottleneck and I wanted to investigate how they could be made less sequential.

Currently reading through Chapter 7 of https://effectivepython.com and made a few areas to explore:
- Item 55 - SAB already does this, but there are some thoughts about restricting the max Queue size - SAB achieves this with busy work querying queue.qsize which I found slow in profiling plus it's inaccurate anyway.
- Item 73 - I'm wondering if heapq could help in assembler to avoid looping through already processed articles, similar for the many functions named get_articles

Generally removing sleeps and waiting for things would probably reduce CPU load.

**Reply by puzzledsab** [reply]
Most of the time there isn't much load but there is some stuff that comes in chunks. Looking at the sleep frequency there doesn't seem to be a noticeable difference between 0.005 and 0.001 seconds. I could just make it another special variable so that normal users could try for them selves.

The queues aren't ideal, particularly the assembler queue. On the other hand they do seem to work much more efficiently than threadpool, at least on my system. I'm just not sure how to implement it in a good way for the receive threads. My only idea so far is to have a request and a result queue and count that we get as many results as requests before continuing.

Having a thread for each connection might work but I think they would have to be smarter than today and not return immediately when data arrives. Otherwise there would be a lot of small results that would need to be handled very frequently. That might complicate things a bit much on the C side.

**Reply by mnightingale** [reply]
I think you misunderstand, I think the use of queues for work is a common and good design, I just have possible issue that they have no `maxsize` set, this means there are a few places in the code which busy loop waiting for the queue size to reduce - which isn't accurate anyway.
If `maxsize` was set the calls to add to the queue would block until there is space available.

**Reply by puzzledsab** [reply]
I was talking about the implementation too, but a different part. I think using a sleep loop instead of maxsize is intentional. It lets the downloader thread break off quickly and do some cleanup if the system is shutting down.

For the assembler queue I don't see any reason for stopping the decoder thread from adding new tasks and I think we'd still need a waiting loop in the downloader thread because `queue.full()` can't wait. It's a very inaccurate measure of how much data needs to be written anyway because one entry can represent anything from 1 to almost 300 parts (with a 4 GB article cache). That's the part I think could be improved.
